### ISSUE TITLE
[Spike] Investigate our Azure setup

### ISSUE THREAD

**Issue by murny** [ualbertalib/jupiter#2570]
> Azure - https://azure.microsoft.com/en-ca/

### Spike Goal
> Document everything we need so developers can setup infrastructure and configure environments for ERA to run in Azure
 
### Spike questions

- Who has access to login into Azure? How can we share access to the team?
- What do we need to do to setup ERA here? (On Kubernetes and using Terraform for provisioning)
- How can we train rest of team to be comfortable and know enough to be able to troubleshoot issues with our Azure environment?
- Get an understanding of what our "Reservations". "Node/Pools" and "Pods/Services" are within Azure.
- Look into security automation tools we have access too for scanning pods/updates in CI, updating of Nodes, and Control Plane updates.
- Look into Encrypted secrets infrastructure that Azure provides
- Document anything that keeps us "vendor locked" to Azure ecosystem and alternatives if we wanted to switch to another provider (AWS/GCP/etc)

**Comment by pgwillia** [comment]
@kgood made a very good point about knowing which decisions we make that "lock" us in to a particular vendor. How easy would it be to move our infrastructure to another host?

**Comment by murny** [comment]
Yeah great point 👍 . I think this is why we want to stay as close to kubernetes and terraform. (alot of providers provide their own provision tools to setup a server for kubernetes or to run kubernete pods, but these vendor lock you into these platforms).

I'll add a note to this ticket for keeping this on our radar when we start diving into Azure and document anything that might lock us to their ecosystem.

**Comment by jefferya** [comment]
In the context of my other position, Persistent Volume Claims (PVC) / storage was an area flagged as needing a different configuration depending on cloud provider: AWS k8s and ComputeCanada/LINCS/Rancher k8s.

**Comment by murny** [comment]
### Spike questions and answers
> Who has access to login into Azure? How can we share access to the team?

See Tricia's comment on getting access here: https://github.com/ualbertalib/jupiter/issues/2570#issuecomment-952182607

> What do we need to do to setup ERA here? (On Kubernetes and using Terraform for provisioning)

Once you have access to our teams Azure Portal, you will need to install and setup Azure CLI:
`brew update && brew install azure-cli`

Then proceed to login into Azure CLI:
`az login`

Terraform will now need the following info:

- "Azure Subscription ID" 
- "Azure Tenant ID"
- "Azure Client ID"
- "Azure Client Secret"

The first two (Subscription ID/Tenant ID) can be retrieved from `az account show` where Subscription ID is the "id" and Tenant ID is the "tenantId" from the command output

For the last two we will need to [setup an azure service principal account](https://docs.microsoft.com/en-us/cli/azure/create-an-azure-service-principal-azure-cli). This can be done with this command:

```
az ad sp create-for-rbac \
  --role="Contributor" \
  --scopes="/subscriptions/{{SUBSCRIPTION ID FROM ABOVE GOES HERE}}"
```
Upon successful completion of running this command it will display the appId, password, which will be used the for the - "Azure Client ID" and "Azure Client Secret". Make sure you keep these credentials in a safe place!

Now that we have all this info, we can set the `terraform.tfvars` with them. Now we can run terraform to provision Jupiter in Azure!

> How can we train rest of team to be comfortable and know enough to be able to troubleshoot issues with our Azure environment?

This will require the team members to be exposed and play around with these tools and environment. Think the best way is to work toward converting our current UAT environment into a cloud environment in Azure which will be provisioned by Terraform and running Kubernetes. This will force our team to learn the tools and get comfortable working with them. Once we have the confidence we can start working toward replacing our current Production stack with the same Cloud stack that UAT will be running.

> Look into Encrypted secrets infrastructure that Azure provides
Haven't looked into it. But essentially the team will need a way to share passwords around (E.g if we let rails do everything and we use credentials, then we need a way to share Master Key around to all devs. If we let terraform or kubernetes do it? Then we need a way to share this password/encryption key around. So basically need something like 1Password. So maybe this is where [Azure's Key Vault](https://azure.microsoft.com/en-us/services/key-vault) can come into play and help us out. Worse case we can just do status quo of what we currently do today (Slack/Private github/emails/etc/etc but something like Key vault or 1password would obviously be a better solution).

> Document anything that keeps us "vendor locked" to Azure ecosystem and alternatives if we wanted to switch to another provider (AWS/GCP/etc)

Honestly not much? Most of the providers will provide similar things (maybe for varying cost differences). We will have to switch our terraform scripts to use the new providers plugin if we decide to switch. This will change various things for setting up and provisioning our infrastructure. So there will be work for switching between providers. But nothing will "vendor lock" us to one particular provider that I can think of? All providers will do similar things at the end of the day (most will provide the kubernentes "control plane" and horizontal scaling/etc for you.

**Comment by pgwillia** [comment]
Here's the Azure CLI installation if you're not on Mac: https://docs.microsoft.com/en-us/cli/azure/install-azure-cli
